File: modules/paycode_events.py
import streamlit as st
import pandas as pd
import requests
import io
import re
from datetime import datetime

# ======================================================
# DATE NORMALIZATION
# ======================================================
def normalize_yyyy_mm_dd(value):
    if not value:
        return None

    if hasattr(value, "strftime"):
        return value.strftime("%Y-%m-%d")

    value = str(value).strip()
    if " " in value:
        value = value.split(" ")[0]

    try:
        datetime.strptime(value, "%Y-%m-%d")
        return value
    except ValueError:
        return None


# ======================================================
# MAIN UI
# ======================================================
def paycode_events_ui():
    st.header("🧾 Paycode Events")

    HOST = st.session_state.HOST.rstrip("/")
    BASE_URL = f"{HOST}/resource-server/api/paycode_events"
    PAYCODES_URL = f"{HOST}/resource-server/api/paycodes"

    headers = {
        "Authorization": f"Bearer {st.session_state.token}",
        "Content-Type": "application/json",
        "Accept": "application/json"
    }

    # ==================================================
    # DOWNLOAD TEMPLATE
    # ==================================================
    st.subheader("📥 Download Upload Template")

    template_df = pd.DataFrame(columns=[
        "id",
        "Paycode Event Name",
        "Description",
        "paycode_id",
        "holiday_name",
        "holiday_date(YYYY-MM-DD)",
        "repeatWeek",
        "repeatWeekday"
    ])

    if st.button("⬇️ Download Template"):
        r = requests.get(PAYCODES_URL, headers=headers)
        paycodes_df = (
            pd.DataFrame(r.json())[["id", "code", "description"]]
            if r.status_code == 200 else pd.DataFrame()
        )

        output = io.BytesIO()
        with pd.ExcelWriter(output, engine="openpyxl") as writer:
            template_df.to_excel(writer, index=False, sheet_name="Paycode Events")
            paycodes_df.to_excel(writer, index=False, sheet_name="Paycodes")

        st.download_button(
            "⬇️ Download Excel",
            output.getvalue(),
            "paycode_events_template.xlsx"
        )

    # ==================================================
    # UPLOAD
    # ==================================================
    st.subheader("📤 Upload Paycode Events")

    uploaded_file = st.file_uploader("Upload CSV / Excel", ["csv", "xlsx", "xls"])
    if not uploaded_file:
        return

    df = (
        pd.read_csv(uploaded_file)
        if uploaded_file.name.endswith(".csv")
        else pd.read_excel(uploaded_file)
    ).fillna("")

    store = {}
    errors = []

    for row_no, row in df.iterrows():
        raw_id = str(row.get("id", "")).strip()
        name = str(row.get("Paycode Event Name", "")).strip()
        description = str(row.get("Description", "")).strip() or name
        paycode_id = str(row.get("paycode_id", "")).strip()
        holiday_name = str(row.get("holiday_name", "")).strip()
        holiday_raw = row.get("holiday_date(YYYY-MM-DD)", "")

        repeat_week = str(row.get("repeatWeek", "")).strip() or "*"
        repeat_weekday = str(row.get("repeatWeekday", "")).strip() or "*"

        if not name or not holiday_name or not holiday_raw or not paycode_id:
            errors.append(f"Row {row_no+1}: Missing mandatory fields")
            continue

        holiday_date = normalize_yyyy_mm_dd(holiday_raw)
        if not holiday_date:
            errors.append(f"Row {row_no+1}: Invalid date {holiday_raw}")
            continue

        year, month, day = map(int, holiday_date.split("-"))

        # 🔐 FIX: ID-FIRST GROUPING (DO NOT CHANGE THIS)
        key = f"ID_{raw_id}" if raw_id.isdigit() else f"NEW_{name}"

        if key not in store:
            base = {
                "name": name,
                "description": description,
                "paycode": {"id": int(float(paycode_id))},
                "schedules": []
            }

            if raw_id.isdigit():
                base["id"] = int(raw_id)

            store[key] = base

        store[key]["schedules"].append({
            "name": holiday_name,
            "startDate": st.session_state.get("START_DATE", "2026-01-01"),
            "repeatYear": year,
            "repeatMonth": month,
            "repeatDay": day,
            "repeatWeek": repeat_week,
            "repeatWeekday": repeat_weekday
        })

    st.session_state.final_body = list(store.values())

    if errors:
        st.error("❌ Some rows were skipped")
        st.text("\n".join(errors))

    st.success(f"Loaded {len(store)} Paycode Events")

    # ==================================================
    # CREATE / UPDATE
    # ==================================================
    st.subheader("🚀 Create / Update Paycode Events")

    if st.button("Submit Paycode Events"):
        results = []

        for payload in st.session_state.final_body:
            is_update = isinstance(payload.get("id"), int)

            if is_update:
                r = requests.put(
                    f"{BASE_URL}/{payload['id']}",
                    headers=headers,
                    json=payload
                )
            else:
                r = requests.post(BASE_URL, headers=headers, json=payload)

            results.append({
                "Paycode Event": payload["name"],
                "Action": "UPDATE" if is_update else "CREATE",
                "HTTP Status": r.status_code,
                "Status": "Success" if r.status_code in (200, 201) else "Failed"
            })

        st.dataframe(pd.DataFrame(results), use_container_width=True)

    # ==================================================
    # DELETE (UNCHANGED)
    # ==================================================
    st.subheader("🗑️ Delete Paycode Events")

    ids_input = st.text_input("Enter Paycode Event IDs (comma-separated)")
    if st.button("Delete Paycode Events"):
        ids = [i.strip() for i in ids_input.split(",") if i.strip().isdigit()]
        for pid in ids:
            r = requests.delete(f"{BASE_URL}/{pid}", headers=headers)
            if r.status_code in (200, 204):
                st.success(f"Deleted Paycode Event ID {pid}")
            else:
                st.error(f"Failed to delete ID {pid} → {r.text}")

    # ==================================================
    # DOWNLOAD EXISTING (UNCHANGED)
    # ==================================================
    st.subheader("⬇️ Download Existing Paycode Events")

    if st.button("Download Existing Paycode Events"):
        r = requests.get(BASE_URL, headers=headers)

        if r.status_code != 200:
            st.error("❌ Failed to fetch Paycode Events")
            return

        rows = []
        for e in r.json():
            for s in e.get("schedules", []):
                ry, rm, rd = s.get("repeatYear"), s.get("repeatMonth"), s.get("repeatDay")
                holiday_date = (
                    f"{int(ry):04d}-{int(rm):02d}-{int(rd):02d}"
                    if str(ry).isdigit() else ""
                )

                rows.append({
                    "id": e.get("id"),
                    "Paycode Event Name": e.get("name"),
                    "Description": e.get("description"),
                    "paycode_id": e.get("paycode", {}).get("id"),
                    "holiday_name": s.get("name"),
                    "holiday_date(YYYY-MM-DD)": holiday_date,
                    "repeatWeek": s.get("repeatWeek", "*"),
                    "repeatWeekday": s.get("repeatWeekday", "*")
                })

        df = pd.DataFrame(rows)
        st.download_button(
            "⬇️ Download CSV",
            df.to_csv(index=False),
            "paycode_events_export.csv"
        )

File: renderer/paycode-events.js
/* =========================================================
 * paycode-events.js — BeeForce Configuration Portal
 * Plain script body. Executed as:
 *   new Function('AppCtx', <this file> + '\nreturn (typeof init==="function")?init(AppCtx):undefined;')
 * Defines a top-level `function init(ctx)` — nothing else is required.
 * No import/export syntax. ctx === AppCtx.
 * ========================================================= */
// =========================================================
// modules/paycode-events.js — BeeForce Configuration Portal
// Owns: screen-pe, screen-pe-create, screen-pe-view, screen-pe-delete
// GET    /api/attendance/paycode_events
// POST   /api/attendance/paycode_events            (create)
// PUT    /api/attendance/paycode_events/{id}        (update)
// DELETE /api/attendance/paycode_events/{id}
// Exports: init(context)
// =========================================================

function init(ctx) {
  const { AppState, apiH, downloadExcel, wireFileInput, parseApiError, Notifications, Router, Audit, ApiTracker, fetchPaycodesRef, modal } = ctx;
  const BASE = () => `${AppState.BASE_URL}/api/attendance/paycode_events`;

  if (!document.getElementById("screen-pe")) {
    const main = document.createElement("div");
    main.id = "screen-pe"; main.style.display = "none";
    main.innerHTML = `
      <div class="bf-breadcrumb"><span class="bc-link" data-go="screen-hub">🏠 Home</span><span class="bc-sep">›</span><span class="bc-cur">📊 Paycode Events</span></div>
      <div class="bf-module-header"><div class="mh-icon">📊</div><div><div class="mh-title">Paycode Events</div><div class="mh-sub">Create · Update · View · Delete</div></div></div>
      <div class="bf-action-grid">
        <div class="bf-action-card" id="pe-action-tpl"><div class="ac-icon">⬇️</div><div class="ac-name">Download Template</div></div>
        <div class="bf-action-card" id="pe-action-create"><div class="ac-icon">📤</div><div class="ac-name">Upload & Create/Update</div></div>
        <div class="bf-action-card" id="pe-action-view"><div class="ac-icon">📋</div><div class="ac-name">View Existing</div></div>
        <div class="bf-action-card" id="pe-action-delete"><div class="ac-icon">🗑️</div><div class="ac-name">Delete</div></div>
      </div>`;
    modal.appendChild(main);

    const create = document.createElement("div");
    create.id = "screen-pe-create"; create.style.display = "none";
    create.innerHTML = `
      <div class="bf-breadcrumb"><span class="bc-link" data-go="screen-hub">🏠 Home</span><span class="bc-sep">›</span><span class="bc-link" data-go="screen-pe">📊 Paycode Events</span><span class="bc-sep">›</span><span class="bc-cur">Create/Update</span></div>
      <div class="bf-info-box">Leave <b>id</b> blank to create, fill to update.</div>
      <div id="pe-create-status" class="bf-status"></div>
      <label class="bf-file-label" for="pe-file"><span id="pe-file-name">📂 Click to select filled template (.xlsx)</span></label>
      <input type="file" id="pe-file" class="bf-file-input" accept=".xlsx,.xls,.csv" />
      <div class="bf-progress" id="pe-c-prog"><div class="bf-pb" id="pe-c-bar" style="width:0%"></div></div>
      <div class="bf-log" id="pe-create-log"></div>
      <button class="bf-btn bf-success" id="pe-submit" disabled>🚀 Submit Paycode Events</button>`;
    modal.appendChild(create);

    const view = document.createElement("div");
    view.id = "screen-pe-view"; view.style.display = "none";
    view.innerHTML = `
      <div class="bf-breadcrumb"><span class="bc-link" data-go="screen-hub">🏠 Home</span><span class="bc-sep">›</span><span class="bc-link" data-go="screen-pe">📊 Paycode Events</span><span class="bc-sep">›</span><span class="bc-cur">View</span></div>
      <div id="pe-view-status" class="bf-status"></div>
      <div class="bf-progress" id="pe-v-prog" style="display:none"><div class="bf-pb" id="pe-v-bar" style="width:0%"></div></div>
      <button class="bf-btn bf-primary" id="pe-fetch">⬇️ Download Paycode Events</button>`;
    modal.appendChild(view);

    const del = document.createElement("div");
    del.id = "screen-pe-delete"; del.style.display = "none";
    del.innerHTML = `
      <div class="bf-breadcrumb"><span class="bc-link" data-go="screen-hub">🏠 Home</span><span class="bc-sep">›</span><span class="bc-link" data-go="screen-pe">📊 Paycode Events</span><span class="bc-sep">›</span><span class="bc-cur">Delete</span></div>
      <div class="bf-warn-box">⚠️ Deletion is permanent and cannot be undone.</div>
      <div id="pe-delete-status" class="bf-status"></div>
      <label class="bf-label">IDs to delete (comma-separated)</label>
      <input type="text" id="pe-del-ids" placeholder="101, 102, 103" />
      <label class="bf-label" style="display:flex;align-items:center;gap:8px;cursor:pointer;font-weight:400;"><input type="checkbox" id="pe-confirm" style="width:auto;margin:0;" />I understand this is permanent</label>
      <div class="bf-log" id="pe-delete-log"></div>
      <button class="bf-btn bf-danger" id="pe-do-delete" style="margin-top:10px;">🗑️ Delete</button>`;
    modal.appendChild(del);

    Router.register("screen-pe", "screen-pe-create", "screen-pe-view", "screen-pe-delete");
    [main, create, view, del].forEach(el => {
      Router.wireDataGoLinks(el);
      el.querySelectorAll("input,textarea").forEach(inp =>
        ["keydown", "keyup", "keypress", "input"].forEach(evt => inp.addEventListener(evt, e => e.stopPropagation(), true))
      );
    });
  }

  document.getElementById("pe-action-tpl").onclick = async () => {
    const btn = document.getElementById("pe-action-tpl"); btn.style.opacity = "0.6";
    const pcs = await fetchPaycodesRef();
    downloadExcel("paycode_events_template.xlsx", [
      { name: "Paycode_Events", tabColor: "1D4ED8",
        headers: ["id", "Paycode Event Name", "Description", "paycode_id", "holiday_name", "holiday_date(YYYY-MM-DD)", "repeatWeek", "repeatWeekday"],
        rows: [], highlightCols: [1, 3, 4, 5] },
      { name: "Paycodes_Master", tabColor: "059669", headers: ["id", "code", "description"], rows: pcs.map(p => [p.id, p.code || "", p.description || ""]) }
    ]);
    btn.style.opacity = "1";
  };
  document.getElementById("pe-action-create").onclick = () => {
    Notifications.clearLog("pe-create-log"); AppState.cache.pe = null;
    document.getElementById("pe-submit").disabled = true;
    document.getElementById("pe-file-name").textContent = "📂 Click to select filled template (.xlsx)";
    Router.go("screen-pe-create");
  };
  document.getElementById("pe-action-view").onclick = () => Router.go("screen-pe-view");
  document.getElementById("pe-action-delete").onclick = () => { Notifications.clearLog("pe-delete-log"); Router.go("screen-pe-delete"); };
  wireFileInput("pe-file", "pe-file-name", "pe-submit", "pe");

  document.getElementById("pe-submit").onclick = async () => {
    const btn = document.getElementById("pe-submit");
    const rows = AppState.cache.pe || [];
    Notifications.clearLog("pe-create-log");
    if (!rows.length) { Notifications.error("pe-create-status", "❌ No data found"); return; }

    const store = {};
    for (const row of rows) {
      const rawId = String(row["id"] || "").trim(), name = String(row["Paycode Event Name"] || "").trim();
      const desc = String(row["Description"] || "").trim() || name, pcId = String(row["paycode_id"] || "").trim();
      const hName = String(row["holiday_name"] || "").trim(), hDate = String(row["holiday_date(YYYY-MM-DD)"] || "").trim();
      const rW = String(row["repeatWeek"] || "*").trim() || "*", rWd = String(row["repeatWeekday"] || "*").trim() || "*";
      if (!name || !pcId || !hName || !hDate) { Notifications.log("pe-create-log", `⚠️ Skipped: ${name || "(no name)"}`, "fail"); continue; }
      const parts = hDate.split("-").map(Number);
      if (parts.length !== 3 || parts.some(isNaN)) { Notifications.log("pe-create-log", `⚠️ Bad date: ${hDate}`, "fail"); continue; }
      const [yr, mo, dy] = parts; const key = rawId || name;
      if (!store[key]) { const base = { name, description: desc, paycode: { id: parseInt(pcId) }, schedules: [] }; if (rawId && !isNaN(rawId)) base.id = parseInt(rawId); store[key] = base; }
      store[key].schedules.push({ name: hName, startDate: "2026-01-01", repeatDay: dy, repeatMonth: mo, repeatYear: yr, repeatWeek: rW, repeatWeekday: rWd });
    }
    const payloads = Object.values(store);
    if (!payloads.length) { Notifications.error("pe-create-status", "❌ No valid rows"); return; }

    Notifications.loading(btn, "Submitting...");
    Notifications.progress.show("pe-c-prog");
    let s = 0, f = 0;
    for (let i = 0; i < payloads.length; i++) {
      const p = payloads[i], isU = typeof p.id === "number";
      try {
        const r = await ApiTracker.fetch(isU ? `${BASE()}/${p.id}` : `${BASE()}`, { method: isU ? "PUT" : "POST", headers: apiH(), body: JSON.stringify(p) }, "Paycode Events");
        if (r.ok) { Notifications.log("pe-create-log", `✅ ${isU ? "Updated" : "Created"}: ${p.name}`, "ok"); s++; Audit.onDbOp(isU ? "updated" : "created", 1); }
        else { Notifications.log("pe-create-log", `❌ Failed: ${p.name} — ${parseApiError(await r.text())}`, "fail"); f++; }
      } catch (e) { Notifications.log("pe-create-log", `❌ Error: ${p.name} — ${e.message}`, "fail"); f++; }
      Notifications.progress.set("pe-c-bar", ((i + 1) / payloads.length) * 100);
    }
    Notifications.log("pe-create-log", `━━━━━━━━━━━━━━━━━━`, "info");
    Notifications.log("pe-create-log", `✅ Success: ${s}  ❌ Failed: ${f}`, f === 0 ? "ok" : "info");
    Notifications.status("pe-create-status", `Done — ${s} success, ${f} failed`, f === 0 ? "success" : "info");
    Notifications.doneLoading(btn, "🚀 Submit Paycode Events");
    Notifications.progress.hide("pe-c-prog");
  };

  document.getElementById("pe-do-delete").onclick = async () => {
    if (!document.getElementById("pe-confirm").checked) { Notifications.error("pe-delete-status", "⚠️ Check confirmation box"); return; }
    const ids = document.getElementById("pe-del-ids").value.split(",").map(s => s.trim()).filter(s => !isNaN(s) && s !== "");
    if (!ids.length) { Notifications.error("pe-delete-status", "❌ No valid IDs"); return; }
    const btn = document.getElementById("pe-do-delete");
    Notifications.loading(btn, "Deleting..."); Notifications.clearLog("pe-delete-log");
    let s = 0, f = 0;
    for (const id of ids) {
      try {
        const r = await ApiTracker.fetch(`${BASE()}/${id}`, { method: "DELETE", headers: apiH() }, "Paycode Events");
        if (r.status === 200 || r.status === 204) { Notifications.log("pe-delete-log", `✅ Deleted: ${id}`, "ok"); s++; }
        else { Notifications.log("pe-delete-log", `❌ Failed: ${id}`, "fail"); f++; }
      } catch (e) { Notifications.log("pe-delete-log", `❌ Error: ${id}`, "fail"); f++; }
    }
    Audit.onDbOp("deleted", s);
    Notifications.status("pe-delete-status", `Done — ${s} deleted, ${f} failed`, f === 0 ? "success" : "info");
    Notifications.doneLoading(btn, "🗑️ Delete");
  };

  document.getElementById("pe-fetch").onclick = async () => {
    const btn = document.getElementById("pe-fetch");
    Notifications.loading(btn, "Connecting to server...");
    Notifications.progress.show("pe-v-prog");
    try {
      Notifications.info("pe-view-status", "⏳ Fetching records..."); Notifications.progress.set("pe-v-bar", 25);
      const r = await ApiTracker.fetch(`${BASE()}`, { headers: apiH() }, "Paycode Events");
      if (!r.ok) throw new Error(`HTTP ${r.status}: ${parseApiError(await r.text())}`);
      AppState.cache.peView = await r.json();
      Notifications.info("pe-view-status", `⏳ Processing ${AppState.cache.peView.length} record(s)...`); Notifications.progress.set("pe-v-bar", 60);
      const rows = [];
      (AppState.cache.peView || []).forEach(e => (e.schedules || []).forEach(s => rows.push([
        e.id, e.name, e.description, e.paycode?.id, s.name,
        `${s.repeatYear}-${String(s.repeatMonth).padStart(2, "0")}-${String(s.repeatDay).padStart(2, "0")}`,
        s.repeatWeek, s.repeatWeekday
      ])));
      if (!rows.length) { Notifications.info("pe-view-status", "ℹ️ No records available to download."); return; }
      Notifications.info("pe-view-status", "⏳ Preparing Excel..."); Notifications.progress.set("pe-v-bar", 85);
      downloadExcel("paycode_events_existing.xlsx", [{
        name: "Paycode_Events", tabColor: "1D4ED8",
        headers: ["id", "Paycode Event Name", "Description", "paycode_id", "holiday_name", "holiday_date(YYYY-MM-DD)", "repeatWeek", "repeatWeekday"],
        rows
      }]);
      Notifications.progress.set("pe-v-bar", 100);
      Notifications.success("pe-view-status", `✅ Download complete — ${AppState.cache.peView.length} record(s)`);
    } catch (e) {
      Notifications.error("pe-view-status", `❌ ${e.message}`); Notifications.progress.set("pe-v-bar", 0);
    } finally {
      Notifications.doneLoading(btn, "⬇️ Download Paycode Events");
      Notifications.progress.hide("pe-v-prog", 2000);
    }
  };
}
